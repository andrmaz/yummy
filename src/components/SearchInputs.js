import React from 'react';
import PropTypes from 'prop-types';
import { useForm, Controller } from 'react-hook-form';
import { 
    Center,
    InputGroup, 
    Input, 
    InputRightElement, 
    Button, 
    Flex, 
    FormControl, 
    Select,
    FormHelperText,
    useColorMode,
} from '@chakra-ui/react';
// import dotenv configuration file
import '../config';

const app_id = process.env.REACT_APP_ID;
const app_key = process.env.REACT_APP_KEY;

const SearchInput = ({ handleChange }) => {

    // To manage color mode chakra exposes the useColorMode
    const { colorMode } = useColorMode();

    const { control, handleSubmit, watch, errors } = useForm({
		mode: 'onChange',
        reValidateMode: 'onChange',
        defaultValues: {
            input: '',
            to: 4,
            diet: 'balanced',
            health: 'tree-nut-free',
        }
    });

    const { input, to, diet, health }  = watch();

    const onSubmit = async(e) => {
        e.preventDefault();
        try {
            const uri = `https://api.edamam.com/search?q=${input}&app_id=${app_id}&app_key=${app_key}&from=0&to=${to}&diet=${diet}&health=${health}`;
            const response = await fetch(uri, {
                method: 'GET',
                headers: {
                    'Content-Encoding': 'gzip',
                    'Cache-Control': 'no-cache'
                }
            });
            const data = await response.json();
            if (typeof data !== 'string'){
                handleChange(data);
            }
            console.log(data);
        } catch (error) {
            throw new Error(error.message);
        }
    };

    return (
        <Center h="20vh" color={colorMode === 'light' ? 'black' : 'white'} minW='300px' maxW='500px' m='auto'>
            <FormControl onSubmit={handleSubmit(onSubmit)}>
                <InputGroup size='md'>
                    <Controller
                        name="input"
                        control={control}
                        rules={{ required: true }}
                        defaultValue=""
                        render={({ onChange, value }) => 
                            <Input 
                                id='input'
                                onChange={onChange} 
                                value={value} 
                                name='input'
                                pr='4.5rem'
                                type='text'
                                variant='outline'
                                placeholder='Enter an ingredient'
                                isRequired={true}
                                errorBorderColor ='red.500'
                                isInvalid={errors.input ? true : false}
                            />
                        }
                    />
                    <InputRightElement width='4.5rem'>
                        <Button 
                            h='1.75rem' 
                            size='sm'
                            variant='solid'
                            colorScheme='blue'
                            isDisabled={input.length < 2 ? true : false}
                            onClick={onSubmit}
                        >
                            Search
                        </Button>
                    </InputRightElement>
                </InputGroup>
                <FormHelperText textAlign='center' fontStyle='oblique' mb='0.5rem'>
                    Select the number of recipies, the diet and the type of meal
                </FormHelperText>
                <Flex justify='space-around' align='center'>
                    <Controller
                        name="to"
                        control={control}
                        defaultValue={4}
                        render={({ onChange, value }) => 
                            <Select 
                                id='to'
                                onChange={onChange} 
                                value={value} 
                                name='to'
                                size='xs' 
                                w='200px'
                                focusBorderColor='blue.500'
                            >
                                <option label={2} value={2} />
                                <option label={3} value={3} />
                                <option label={4} value={4} />
                                <option label={5} value={5} />
                                <option label={6} value={6} />
                            </Select>
                        }
                    />
                    <Controller
                        name="diet"
                        control={control}
                        defaultValue='balanced'
                        render={({ onChange, value }) => 
                        <Select 
                            id='diet'
                            onChange={onChange} 
                            value={value} 
                            name='diet'
                            size='xs' 
                            w='200px'
                            focusBorderColor='blue.500'
                        >
                            <option label='balanced' value='balanced' />
                            <option label='high-protein' value='high-protein' />
                            <option label='high-fiber' value='high-fiber' />
                            <option label='low-fat' value='low-fat' />
                            <option label='low-carb' value='low-carb' />
                        </Select>
                        }
                    />
                    <Controller
                        name="health"
                        control={control}
                        defaultValue='tree-nut-free'
                        render={({ onChange, value }) => 
                        <Select 
                            id='health'
                            onChange={onChange} 
                            value={value} 
                            name='health'
                            size='xs' 
                            w='200px'
                            focusBorderColor='blue.500'
                        >
                            <option label='tree-nut-free' value='tree-nut-free' />
                            <option label='peanut-free' value='peanut-free' />
                            <option label='vegetarian' value='vegetarian' />
                        </Select>
                        }
                    />
                </Flex>
            </FormControl>
        </Center>
    );
};

SearchInput.propTypes = {
    handleChange: PropTypes.func,
};

export default SearchInput;
