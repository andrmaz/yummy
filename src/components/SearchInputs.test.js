import React from 'react';
import { shallow } from 'enzyme';
import SearchInputs from './SearchInputs';

let wrapper;
const mockSearch = jest.fn();
const mockValues = {
    diet: 'balanced',
    health: 'tree-nut-free',
    input: '',
    to: 4,
};

describe('SearchInput Component', () => {
    beforeEach(() => {
        wrapper = shallow(<SearchInputs onSearch={mockSearch} />);
        expect.hasAssertions();
    });
    it('should render Component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    it('should have input value as required', () => {
        const element = wrapper.find('[data-testid="a"]');
        expect(element.prop('rules')).toHaveProperty('required', true);
    });
    it('should have default input as empty string', () => {
        const element = wrapper.find('[data-testid="a"]');
        expect(element.prop('defaultValue')).toEqual('');
    });
    it('should have default to value set to four', () => {
        const element = wrapper.find('[data-testid="b"]');
        expect(element.prop('defaultValue')).toEqual(4);
    });
    it('should have default diet value set to balanced', () => {
        const element = wrapper.find('[data-testid="c"]');
        expect(element.prop('defaultValue')).toEqual('balanced');
    });
    it('should have default health value set to tree-nut-free', () => {
        const element = wrapper.find('[data-testid="d"]');
        expect(element.prop('defaultValue')).toEqual('tree-nut-free');
    });
    it('should have current values set to default values', () => {
        const element = wrapper.find('[data-testid="a"]');
        expect(element.prop('control')).toHaveProperty('defaultValuesRef.current', mockValues);
    });
    it('should not click button on empty input', () => {
        const element = wrapper.find('[id="button"]');
        element.simulate('click');
        expect(mockSearch).toBeCalledTimes(0);
    });
});
