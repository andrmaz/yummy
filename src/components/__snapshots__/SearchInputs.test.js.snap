// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput Component should render Component correctly 1`] = `
<Center
  color="white"
  h="20vh"
  m="auto"
  maxW="500px"
  minW="300px"
>
  <FormControl
    onSubmit={[Function]}
  >
    <InputGroup
      size="md"
    >
      <Controller
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
        data-testid="a"
        defaultValue=""
        name="input"
        render={[Function]}
        rules={
          Object {
            "required": true,
          }
        }
      />
      <InputRightElement
        width="4.5rem"
      >
        <Button
          colorScheme="blue"
          h="1.75rem"
          id="button"
          isDisabled={true}
          onClick={[Function]}
          size="sm"
          variant="solid"
        >
          Search
        </Button>
      </InputRightElement>
    </InputGroup>
    <FormHelperText
      fontStyle="oblique"
      mb="0.5rem"
      textAlign="center"
    >
      Select the number of recipies, the diet and the type of meal
    </FormHelperText>
    <Flex
      align="center"
      justify="space-around"
    >
      <Controller
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
        data-testid="b"
        defaultValue={4}
        name="to"
        render={[Function]}
      />
      <Controller
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
        data-testid="c"
        defaultValue="balanced"
        name="diet"
        render={[Function]}
      />
      <Controller
        control={
          Object {
            "defaultValuesRef": Object {
              "current": Object {
                "diet": "balanced",
                "health": "tree-nut-free",
                "input": "",
                "to": 4,
              },
            },
            "fieldArrayDefaultValuesRef": Object {
              "current": Object {},
            },
            "fieldArrayNamesRef": Object {
              "current": Set {},
            },
            "fieldArrayValuesRef": Object {
              "current": Object {},
            },
            "fieldsRef": Object {
              "current": Object {},
            },
            "fieldsWithValidationRef": Object {
              "current": Object {},
            },
            "formState": Object {
              "dirtyFields": Object {},
              "errors": Object {},
              "isDirty": false,
              "isSubmitSuccessful": false,
              "isSubmitted": false,
              "isSubmitting": false,
              "isValid": true,
              "isValidating": false,
              "submitCount": 0,
              "touched": Object {},
            },
            "formStateRef": Object {
              "current": Object {
                "dirtyFields": Object {},
                "errors": Object {},
                "isDirty": false,
                "isSubmitSuccessful": false,
                "isSubmitted": false,
                "isSubmitting": false,
                "isValid": true,
                "isValidating": false,
                "submitCount": 0,
                "touched": Object {},
              },
            },
            "getValues": [Function],
            "isFormDirty": [Function],
            "mode": Object {
              "isOnAll": false,
              "isOnBlur": false,
              "isOnChange": true,
              "isOnSubmit": false,
              "isOnTouch": false,
            },
            "reValidateMode": Object {
              "isReValidateOnBlur": false,
              "isReValidateOnChange": true,
            },
            "readFormStateRef": Object {
              "current": Object {
                "constructor": true,
                "dirtyFields": true,
                "errors": true,
                "isDirty": true,
                "isSubmitSuccessful": true,
                "isSubmitted": true,
                "isSubmitting": true,
                "isValid": true,
                "isValidating": true,
                "submitCount": true,
                "touched": true,
              },
            },
            "register": [Function],
            "removeFieldEventListener": [Function],
            "resetFieldArrayFunctionRef": Object {
              "current": Object {},
            },
            "setValue": [Function],
            "shallowFieldsStateRef": Object {
              "current": Object {},
            },
            "shouldUnregister": true,
            "trigger": [Function],
            "unregister": [Function],
            "updateFormState": [Function],
            "updateWatchedValue": [Function],
            "useWatchFieldsRef": Object {
              "current": Object {},
            },
            "useWatchRenderFunctionsRef": Object {
              "current": Object {},
            },
            "validFieldsRef": Object {
              "current": Object {},
            },
            "validateResolver": undefined,
            "watchInternal": [Function],
          }
        }
        data-testid="d"
        defaultValue="tree-nut-free"
        name="health"
        render={[Function]}
      />
    </Flex>
  </FormControl>
</Center>
`;
